### DISCUSSION TITLE
How to model app state of a large list of complex objects?

### DISCUSSION THREAD

**Discussion by rollie42** [pmndrs/zustand#1751] [Q&A]
Right now I have a customer management app, with app state like:

```js
    interface Customer {
      name: string,
      address: Address,
      preferences: Preferences
      ... 10 more fields ...
    }
    
    const customers: Customer[] = get500Customers()
    const store = create ((set) => ({
      customers: customers,
      setCustomers: (newCustomerList: Customer[]) => set((state) => ({ customers: newCustomerList })
    })
```

But I find the updates are somewhat slow to be reflected in the UI. I'm new to zustand so probably I'm not organizing this in the best way possible, so...should the above work, and is the cause of the slowness I'm seeing likely elsewhere? If not, how would you organize the above?  I could imagine breaking the various fields of customer out into states themselves, so 

```js
const customerNames = {}
for (const cust of customers) {
  customerNames[cust.id] = cust.name
}
const customerStore = create ((set) => ({
  names: customerNames,
  setNames: ...
})
```

or, changing the state to be keyed by customer Id, and then have separate state for the list of customers, so:

```js
const state = {}
for (const cust  of customers) {
  state[cust.id] = cust
  state[`set${cust.id}`] = ...setter...
}
const customerStore = create ((set) => (state)
```

but I haven't seen this sort of dynamic store creation recommended anywhere, so I'm doubtful it will work as desired.

Thoughts?

**Comment by rollie42** [comment]
Would love if anyone has some insight into the above!

**Comment by stevebeauge** [comment]
I have the same question. I want to design an app that can create and open multiple tabs, each with a complex UI.
As far as I understand, the zustand paradigm consists in creating a single centralized store and immutably updating its content inside actions.
That means every single store update will lead to duplicating the whole state tree, and its probably what's causing the slowness.
I wonder if Jotai, is better for large scale application. Because atoms are small, independent chunk of data, I imagine the cost of updating the state far lower than zustand.
However, Jotai brings its own challenge, especially when dealing with pushing local atoms under the component tree (have to mis context and atom family I guess).

I'm neither a zustand nor jotai expert. I'd appreciate some insights by ones that known

**Reply by adzm** [reply]
Create a store in your tab component or wherever and put it within a context provider; then just use that store via context. Works great for independent zustand stores.
